File: exercise04-A/src/program.cpp
// Exercise 04-A: Palindromic Permutations

#include <iostream>
#include <string>
#include <unordered_set>

// Prototypes

bool IsPalindromic(const std::string& s);
void GetLettersWithOddCounts(const std::string& s, std::unordered_set<char>& odds);
void FlipMembership(char c, std::unordered_set<char>& s);

// Main execution

int main(int argc, char* argv[]) {
    std::string word;

    while (std::cin >> word)
        std::cout << (IsPalindromic(word) ? "Yes" : "No") << '\n';

    return 0;
}

// Functions

bool IsPalindromic(const std::string& s) {
    std::unordered_set<char> odds;
    GetLettersWithOddCounts(s, odds);
    return odds.size() <= 1;
}

void GetLettersWithOddCounts(const std::string& s, std::unordered_set<char>& odds) {
    for (char c : s) FlipMembership(c, odds);
}

void FlipMembership(char c, std::unordered_set<char>& s) {
    if (!s.insert(c).second) s.erase(c);
}

// vim: set sts=4 sw=4 ts=8 ft=cpp:

File: exercise04-A/src/program.py
#!/usr/bin/env python3

# Exercise 04-A: Palindromic Permutations

import sys

# Functions


def is_palindromic(s: str) -> bool:
    return len(get_letters_with_odd_counts(s)) <= 1


def get_letters_with_odd_counts(s: str) -> set[int]:
    odds = set()
    for char in s:
        flip_membership(char, odds)

    return odds


def flip_membership(element: any, the_set: set):
    if element in the_set:
        the_set.remove(element)
    else:
        the_set.add(element)


# Main Execution


def main() -> None:
    for word in sys.stdin:
        print("Yes" if is_palindromic(word.strip()) else "No")


if __name__ == "__main__":
    main()

# vim: set sts=4 sw=4 ts=8 expandtab ft=python:
